<a href="https://colab.research.google.com/github/putty21/Latihan-Python/blob/main/Latihan1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **A. Errors and Debugging**


### **1. Controlling Exceptions: %xmode**

In [ ]:
def func1(a, b):
  return a / b
def func2(x):
  a = x
  b = x - 1
  return func1(a, b)

In [ ]:
func2(1)

ZeroDivisionError: ignored

Memanggil func2 menghasilkan sebuah error, dengan magic function %xmode IPython bisa mengontrol jumlah informasi yang ditampilkan ketika error muncul. Ada tiga macam %xmode, plain, context, dan verbose.


In [ ]:
%xmode Plain

Exception reporting mode: Plain


In [ ]:
func2(1)

ZeroDivisionError: ignored

In [ ]:
%xmode Verbose

Exception reporting mode: Verbose


In [ ]:
func2(1)

ZeroDivisionError: ignored

### **2. Debugging: When Reading Tracebacks Is Not Enough**

In [ ]:
%debug

> <ipython-input-2-9b01cf0b0f56>(2)func1()
      1 def func1(a, b):
----> 2   return a / b
      3 def func2(x):
      4   a = x
      5   b = x - 1

ipdb> print(a)
1
ipdb> print(b)
0
ipdb> quit



PYDEV DEBUGGER WARNING:
sys.settrace() should not be used when the debugger is being used.
This may cause the debugger to stop working correctly.
If this is needed, please check: 
http://pydev.blogspot.com/2007/06/why-cant-pydev-debugger-work-with.html
to see how to restore the debug tracing back correctly.
Call Location:
  File "/usr/lib/python3.7/bdb.py", line 357, in set_quit
    sys.settrace(None)



In [ ]:
%debug

> <ipython-input-2-9b01cf0b0f56>(2)func1()
      1 def func1(a, b):
----> 2   return a / b
      3 def func2(x):
      4   a = x
      5   b = x - 1

ipdb> up
> <ipython-input-2-9b01cf0b0f56>(6)func2()
      2   return a / b
      3 def func2(x):
      4   a = x
      5   b = x - 1
----> 6   return func1(a, b)

ipdb> print(x)
1
ipdb> up
> <ipython-input-7-7cb498ea7ed1>(1)<module>()
----> 1 func2(1)

ipdb> down
> <ipython-input-2-9b01cf0b0f56>(6)func2()
      2   return a / b
      3 def func2(x):
      4   a = x
      5   b = x - 1
----> 6   return func1(a, b)

ipdb> quit


Ini memungkinkan user dengan cepat mengetahui penyebab kesalahan dan juga fungsi menyebabkan kesalahan. Lalu menggunakan magic function %pbd untuk mengaktifkan debugger secara otomatis ketika terjadi error.

In [ ]:
%xmode Plain
%pdb on
func2(1)

Exception reporting mode: Plain
Automatic pdb calling has been turned ON


ZeroDivisionError: ignored

> <ipython-input-2-9b01cf0b0f56>(2)func1()
      1 def func1(a, b):
----> 2   return a / b
      3 def func2(x):
      4   a = x
      5   b = x - 1

ipdb> print(b)
0
ipdb> quit


# **B. Profiling and Timing Code**

### **1. Timing Code Snippets: %timeit and %time**

In [ ]:
%timeit sum(range(100))

The slowest run took 5.01 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 5: 1.09 µs per loop


%timeit menghitung waktu eksekusi berulang dari sebuah statement untuk akurasi yang lebih tepat.

In [ ]:
%%timeit
total = 0
for i in range(1000):
  for j in range(1000):
    total += i * (-1) ** j

1 loop, best of 5: 333 ms per loop


%%timeit menghitung waktu eksekusi berulang dari kode.

In [ ]:
import random
L = [random.random() for i in range(100000)]
%timeit L.sort()

The slowest run took 32.56 times longer than the fastest. This could mean that an intermediate result is being cached.
1000 loops, best of 5: 739 µs per loop


Menyortir list yang sudah disortir sebelumnya memakan waktu yang lebih cepat daripada menyortir list yang belum di sortir dengan perulangan. Fungsi magic %time merupakan pilihan yang tepat.

In [ ]:
import random
L = [random.random() for i in range(100000)]
print("sorting an unsorted list:")
%time L.sort()

sorting an unsorted list:
CPU times: user 23 ms, sys: 0 ns, total: 23 ms
Wall time: 23.7 ms


In [ ]:
print("sorting an already sorted list:")
%time L.sort()

sorting an already sorted list:
CPU times: user 2.83 ms, sys: 0 ns, total: 2.83 ms
Wall time: 2.91 ms


Berikut terlihat hasil perbedaan waktu ketika menyortir list yang sudah disortir dibandingkan dengan yang belum disortir. Terlihat juga bahwa fungsi magic %timeit memakan waktu yang lebih cepat dibandingkan %time.

Untuk %time dan %timeit, menggunakan double-percent-sign memungkinkan untuk melakukan timing multiline script.

In [ ]:
%%time
total = 0
for i in range(1000):
  for j in range(1000):
    total += i * (-1) ** j

CPU times: user 417 ms, sys: 0 ns, total: 417 ms
Wall time: 422 ms


### **2. Profiling Full Scripts: %prun**

In [ ]:
def sum_of_lists(N):
  total = 0
  for i in range(5):
    L = [j ^ (j >> i) for j in range(N)]
    total += sum(L)
  return total

Memanggil %prun dengan function call untuk menampilkan deskripsi hasil.

In [ ]:
%prun sum_of_lists(1000000)

(hasil terlihat pada tab bantuan)

---

Pada hasil terlihat total waktu pada setiap function call, dimana eksekusi yang banyak memakan waktu.

### **3. Line-by-Line Profiling with %lprun**



In [ ]:
pip install line_profiler

     |████████████████████████████████| 71kB 7.7MB/s 


In [ ]:
%load_ext line_profiler

The line_profiler extension is already loaded. To reload it, use:
  %reload_ext line_profiler


In [ ]:
%lprun -f sum_of_lists sum_of_lists(5000)

Dengan %lprun, deskripsi dari proses dijelaskan per linenya.

### **4. Profiling Memory Use: %memit and %mprun**

In [ ]:
pip install memory_profiler

  Created wheel for memory-profiler: filename=memory_profiler-0.58.0-cp37-none-any.whl size=30180 sha256=a16d8478a7bfdb44271f966fef8e73acf180c4c09a6dbea2eb7fcc1a88b50552
  Stored in directory: /root/.cache/pip/wheels/02/e4/0b/aaab481fc5dd2a4ea59e78bc7231bb6aae7635ca7ee79f8ae5
Successfully built memory-profiler


In [ ]:
%load_ext memory_profiler

The memory_profiler extension is already loaded. To reload it, use:
  %reload_ext memory_profiler


In [ ]:
%memit sum_of_lists(1000000)

peak memory: 247.42 MiB, increment: 57.10 MiB


Ekstensi memory profiler menyediakan dua fungsi magic, yaitu %memit (menyediakan perhitungan memori yang setara dengan %timeit) dan %mprun (menyediakan perhitungan memori yang setara dengan %lprun).

Pada sintaks diatas terlihat bahwa memori yang digunakan sebesar 247 MB.

In [ ]:
%%file mprun_demo.py
def sum_of_lists(N):
  total = 0
  for i in range(5):
    L = [j ^ (j >> i) for j in range(N)]
    total += sum(L)
    del L # remove reference to L
  return total

Writing mprun_demo.py


In [ ]:
from mprun_demo import sum_of_lists
%mprun -f sum_of_lists sum_of_lists(1000000)

(hasil terlihat pada tab bantuan)

---
Pada kolom increment terlihat seberapa besar setiap baris memengaruhi total memori.
